toy-problems/generateParenthesis/generateParenthesis.js

```diff
@@ -0,0 +1,65 @@
+/* generate parenthesis
+given n pairs of parenthesis write
+function to generate all combinttion of well paired
+parenthesis
+
+Example:
+  Input: n = 3
+
+  Output: [
+    "((()))",
+    "(()())",
+    "(())()",
+    "()(())",
+    "()()()"
+  ]
+
+         (
+     (       )
+   (   )   (   )
+
+  '(((' 3
+  '(()' 1
+  '()' 0
+  '())' -1 BAD!!!
+
+  input: number of parens
+  output: is an array of strings(possible paren pairs)
+
+descision tree beccuase of permutations
+
+function generateParens that takes a number n
+  create counter variable set to 0
+  create result an array of valid paren combos
+
+  inner function parenBuilder
+
+  invoke parenBuilder with param
+
+  return result
+
+*/
+
+const generateParenthesis = (n) => {
+  const result = [];
+
+  const parenBuilder = (current, branchCount) => {
+    if (branchCount === 0 && current.length === n * 2) {
+      result.push(current);
+      return;
+    }
+
+    if (current.length > n * 2) return;
+
+    if (branchCount < 0) return;
+
+    parenBuilder(`${current}(`, branchCount + 1);
+    parenBuilder(`${current})`, branchCount - 1);
```

Comment:
Another way this can be handled is to not even invoke parenBuilder if you know it will make branch count go negative.
```javascript
if (branchCount >= 1) {
 parenBuilder(`${current})`, branchCount - 1); 
}
```
Then we know that branchCount will never be negative in the first place, so we can take away line 54.

Reply:
I should clarify, this would only apply to the invocation at line 57, because we can always add an open parenthesis.